lib/streamlit/elements/data_editor.py

```diff
@@ -0,0 +1,582 @@
+        A dictionary of edited cells, where the key is the cell's row and
+        column position (row:column), and the value is the new value of the cell.
+
+    added_rows : List[Dict[str, str | int | float | bool | None]]
+        A list of added rows, where each row is a dictionary of column position
+        and the respective value.
+
+    deleted_rows : List[int]
+        A list of deleted rows, where each row is the numerical position of the deleted row.
+    """
+
+    edited_cells: Dict[str, str | int | float | bool | None]
+    added_rows: List[Dict[str, str | int | float | bool | None]]
+    deleted_rows: List[int]
+
+
+# A mapping of column names/IDs to column configs.
+ColumnConfigMapping: TypeAlias = Dict[Union[int, str], ColumnConfig]
+
+
+def _marshall_column_config(
+    proto: ArrowProto, columns: Optional[Dict[Union[int, str], ColumnConfig]] = None
+) -> None:
+    """Marshall the column config into the proto.
+
+    Parameters
+    ----------
+    proto : ArrowProto
+        The proto to marshall into.
+
+    columns : Optional[ColumnConfigMapping]
+        The column config to marshall.
+    """
+    if columns is None:
+        columns = {}
+
+    # Ignore all None values and prefix columns specified by index
+    def remove_none_values(input_dict: Dict[Any, Any]) -> Dict[Any, Any]:
+        new_dict = {}
+        for key, val in input_dict.items():
+            if isinstance(val, dict):
+                val = remove_none_values(val)
+            if val is not None:
+                new_dict[key] = val
+        return new_dict
+
+    proto.columns = json.dumps(
+        {
+            (f"col:{str(k)}" if isinstance(k, int) else k): v
+            for (k, v) in remove_none_values(columns).items()
+        }
+    )
+
+
+@dataclass
+class DataEditorSerde:
+    """DataEditorSerde is used to serialize and deserialize the data editor state."""
+
+    def deserialize(self, ui_value: Optional[str], widget_id: str = "") -> EditingState:
+        return (  # type: ignore
+            {
+                "edited_cells": {},
+                "added_rows": [],
+                "deleted_rows": [],
+            }
+            if ui_value is None
+            else json.loads(ui_value)
+        )
+
+    def serialize(self, editing_state: EditingState) -> str:
+        return json.dumps(editing_state, default=str)
+
+
+def _parse_value(value: Union[str, int, float, bool, None], dtype) -> Any:
+    """Convert a value to the correct type.
+
+    Parameters
+    ----------
+    value : str | int | float | bool | None
+        The value to convert.
+
+    dtype
+        The type of the value.
+
+    Returns
+    -------
+    The converted value.
+    """
+    if value is None:
+        return None
+
+    # TODO(lukasmasuch): how to deal with date & time columns?
+
+    # Datetime values try to parse the value to datetime:
+    # The value is expected to be a ISO 8601 string
+    if is_datetime64_any_dtype(dtype):
```

Comment:
Should we also use [is_bool_dtype](https://pandas.pydata.org/docs/reference/api/pandas.api.types.is_bool_dtype.html) and [is_string](https://pandas.pydata.org/docs/reference/api/pandas.api.types.is_string_dtype.html)?